Add millisecond assertions for `DateTime` and `DateTimeOffset`

bprussell opened the pull request:
Introduced new assertion methods `HaveMilliseconds` and `NotHaveMilliseconds` for `DateTime` and `DateTimeOffset` classes in `DateTimeAssertions.cs` and `DateTimeOffsetAssertions.cs`. Updated `datetimespans.md` with examples of the new methods. Added test files `DateTimeAssertionSpecs.HaveMilliseconds.cs` and `DateTimeOffsetAssertionSpecs.HaveMilliseconds.cs` to verify the new functionality.

Closes #2738 

<!-- Please provide a description of your changes above the IMPORTANT checklist -->


## IMPORTANT 

* [ ] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [x] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [x] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [x] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [x] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

bprussell commented:
My first time contributing. I know there's not an `api-approved` label on the issue, so I'll leave this as a Draft PR until there is. I also couldn't figure out how to run the build.ps1 script. It failed with the message `Member 'GitVersion' is required to be not null or empty`.

dennisdoomen commented:
Please rebase on top of `develop`

dennisdoomen commented on Src/FluentAssertions/Primitives/DateTimeAssertions.cs:
```diff
@@ -715,6 +715,62 @@ public AndConstraint<TAssertions> NotHaveSecond(int unexpected,
     }
 
+    /// <summary>
+    /// Asserts that the current <see cref="DateTime"/>  has the <paramref name="expected"/> millisecond.
```
⛏️ (nitpick) too many spaces
⛏️ (nitpick) it's "milliseconds"

dennisdoomen commented on Src/FluentAssertions/Primitives/DateTimeAssertions.cs:
```diff
@@ -715,6 +715,62 @@ public AndConstraint<TAssertions> NotHaveSecond(int unexpected,
+    /// <summary>
+    /// Asserts that the current <see cref="DateTime"/>  has the <paramref name="expected"/> millisecond.
+    /// </summary>
+    /// <param name="expected">The expected milliseconds of the current value.</param>
```
⛏️ "The expected number of milliseconds."

dennisdoomen commented on Src/FluentAssertions/Primitives/DateTimeAssertions.cs:
```diff
@@ -715,6 +715,62 @@ public AndConstraint<TAssertions> NotHaveSecond(int unexpected,
+            .ForCondition(Subject.HasValue)
+            .FailWith(", but found a <null> DateTime.")
+            .Then
+            .ForCondition(Subject.Value.Millisecond == expected)
```
🔧 My IDE tells me that `Subject.Value` can be `null`. You can suppress that with `!`

dennisdoomen commented on Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.HaveMillisecond.cs:
```diff
@@ -0,0 +1,102 @@
+    public class HaveMillisecond
+    {
+        [Fact]
+        public void When_asserting_subject_datetime_should_have_milliseconds_with_the_same_value_it_should_succeed()
```
🔧 For a while, we've moved away from `When` `Should` constructs in test names and switch to more fact-based naming. See `MemberMatchingSpecs` for example.

dennisdoomen commented on Src/FluentAssertions/Primitives/DateTimeAssertions.cs:
```diff
@@ -715,6 +715,62 @@ public AndConstraint<TAssertions> NotHaveSecond(int unexpected,
+    /// <param name="becauseArgs">
+    /// Zero or more objects to format using the placeholders in <paramref name="because" />.
+    /// </param>
+    public AndConstraint<TAssertions> HaveMillisecond(int expected,
```
🤔 I feel that `HaveMilliseconds` is more proper English.

bprussell replied:
I thought it would be more consistent with the existing APIs, for example, `HaveMinute` and `HaveSecond`.

dennisdoomen replied:
Yeah, I figured that, but it didn't feel correct. After "consulting" with ChatGPT I'm more convinced that `HaveMilliseconds` is the right name.

bprussell commented:
Made the following changes per the review:

- Renamed methods, comments, etc. to use plural form of milliseconds.
- Renamed unit test methods: I'm not 100% sure I understood the assignment here, so let me know what you think.
- Removed extra spaces in XML documentation: I did this throughout the files I touched.
- Used the null forgiving operator `!.` to suppress IDE warnings about `Subject.Value` potentially being null: I also did this throughout the files I touched. My IDE did not give this warning, so you'll have to let me know if it satisfies yours.

I also was able to successfully run AcceptAPIChanges.ps1, but I am still not able to run build.ps1 as per my previous comment. Also, the issue could use the `api-approved` label.

IT-VBFK commented:
> but I am still not able to run build.ps1

Can you run `build.cmd`?

bprussell commented:
> > but I am still not able to run build.ps1
> 
> Can you run `build.cmd`?

No, it fails with this error:

```
╬══════════════════════
║ Errors & Warnings
╬═════════════
​
[WRN] : Could not inject value for Build.GitVersion
​
═════════════════════════════════════════════════
Target                       Status      Duration
─────────────────────────────────────────────────
InstallNode                  NotRun              
SpellCheck                   NotRun              
Clean                        NotRun              
Restore                      NotRun              
CalculateNugetVersion        NotRun              
Compile                      NotRun              
TestingPlatformFrameworks    NotRun              
VSTestFrameworks             NotRun              
TestFrameworks               NotRun              
UnitTestsNet6OrGreater       NotRun              
UnitTestsNet47               NotRun              
UnitTests                    NotRun              
CodeCoverage                 NotRun              
ApiChecks                    NotRun              
Pack                         NotRun              
Push                         NotRun              
─────────────────────────────────────────────────
Total                                      < 1sec
═════════════════════════════════════════════════
​
Build failed on 10/29/2024 5:56:31 PM. (╯°□°）╯︵ ┻━┻
```

dennisdoomen commented:
> [WRN] : Could not inject value for Build.GitVersion

Are you using HTTPS or SSL for cloning the repo?

ITaluone commented:
> Could not inject value for Build.GitVersion

Huh.. it seems, that you have to install the gitversion.tool: `dotnet tool install GitVersion.Tool --global`

bprussell commented:
I did that, and got the same error. I noticed it installed version 6.0.3, while several places in the project reference version 6.0.2. I uninstalled the tool and reinstalled it with `dotnet tool install GitVersion.Tool --global --version 6.0.2`, and I got the same error.

I am using a supported version of Visual Studio according to the readme. I am not 100% certain that I have the "Build Tools 2022 (including the Universal Windows Platform build tools)" that the readme mentions, because I couldn't find anything by that exact name. I have the following things installed that I thought were relevant. Let me know if there's something specific that is needed that I didn't list here:
- C# and Visual Basic Roslyn compilers
- .NET Framework 4.8 and 4.8.1 SDKs (which should cover the 4.7 requirement)
- .NET 8.0 SDK
- C++ Universal Windows Platform support for v143 build tools (ARM64/ARM64EC)

dennisdoomen commented:
> Huh.. it seems, that you have to install the gitversion.tool: `dotnet tool install GitVersion.Tool --global`

No, you don't. It's automatically installed as a local tool and used by Nuke.

dennisdoomen commented:
> No, it fails with this error:

Your build log should start with some kind of dump from GitVersion.

bprussell commented:
Here is the full output from running build.ps1:

https://gist.github.com/bprussell/a2f0341f170f414a9badfd06a70f2b89

dennisdoomen reviewed (changes requested):
I somehow forgot to review the tests.

bprussell commented:
Updated unit tests based on the feedback. I feel it's worth noting that I had just copied the patterns from the tests for the other `DateTime` parts. These were created about 1 year ago. Perhaps they should be updated at some point too.

dennisdoomen commented:
> Updated unit tests based on the feedback. I feel it's worth noting that I had just copied the patterns from the tests for the other `DateTime` parts. These were created about 1 year ago. Perhaps they should be updated at some point too.

Noted. We're just trying to make sure the new stuff meets our current standards. Thanks for addressing the comments.

BTW. You know you should mark comments you've address as "Resolved". That's how I know which one I can expect to be resolved or not.

ITaluone commented:
> Here is the full output from running build.ps1:
> 
> https://gist.github.com/bprussell/a2f0341f170f414a9badfd06a70f2b89

Ahm.. do you have tags in you git history?

This seems to be the issue, that gitversion cannot determine the base version:
```
GitVersion.GitVersionException: No base versions determined on the current branch.
      at GitVersion.VersionCalculation.NextVersionCalculator.CalculateNextVersion(IBranch branch, IGitVersionConfiguration configuration) in /_/src/GitVersion.Core/VersionCalculation/VersionCalculators/NextVersionCalculator.cs:line 161
      at GitVersion.VersionCalculation.NextVersionCalculator.FindVersion() in /_/src/GitVersion.Core/VersionCalculation/VersionCalculators/NextVersionCalculator.cs:line 53
      at GitVersion.GitVersionCalculateTool.CalculateVersionVariables() in /_/src/GitVersion.Core/Core/GitVersionCalculateTool.cs:line 45
      at GitVersion.GitVersionExecutor.RunGitVersionTool(GitVersionOptions gitVersionOptions) in /_/src/GitVersion.App/GitVersionExecutor.cs:line 66
```

What's the output of: `git log --graph --format="%h %cr %d" --decorate --date=relative --all --remotes=*`

dennisdoomen commented:
> This seems to be the issue, that gitversion cannot determine the base version:

Yeah, looks like he used a special kind of clone that didn't include all the other branches and tags.

ITaluone commented:
`git fetch --all --tags` should fix that..

bprussell commented:
I tried `git fetch --all --tags`, no dice. I tried a fresh clone. Same result.

https://gist.github.com/bprussell/8e96448afc01b4120d1963199cf68add

ITaluone commented:
Does it work, if you delete/rename the `GitVersion.yml`?

bprussell commented:
Yes, I renamed the file and it worked. Is that sufficient to satisfy the checklist?

ITaluone commented:
BTW.. your gist log also mentions `INFO [24-10-31 10:33:23:48] An orphaned branch '2738-assert-milliseconds' has been detected and will be skipped=True.`

I don't know how that could happen, but does this change if you checkout develop and run it?

bprussell commented:
Running the spellcheck against the `develop` branch works. 🤷‍♂️

I really don't think I did anything abnormal when creating the fork/branch, but that was 3 weeks ago, so it's hard to say with certainty.

ITaluone commented:
What about checkout a new branch (e.g. `tmp`) on where this PR branch points to?

bprussell commented:
Yeah, that's what I did with `pr/2738-assert-milliseconds`; same bad result. I just tried it again with `tmp` like you said; same result.

ITaluone commented:
And: rebase to current develop, maybe this fixes some issues as well...

bprussell commented:
Rebased, which needed to happen anyway, but I'm still getting the GitVersion error during the build script, unless I rename the file.

ITaluone commented:
I am running out of ideas :D 

Can you try to delete content from `GitVersion.yml` and see if at some point it is working?

ITaluone commented:
Another question, as I see that you have no master branch in your fork: Do you have [shallow-forked](https://github.com/orgs/community/discussions/13128) this repo?

bprussell commented:
I tried removing various lines from the GitVersion file, including leaving only the `workflow:` line. Nothing worked, except when I made it an empty file.

I believe I chose the default option when forking the repo, which is to fork the `develop` branch only.

![image](https://github.com/user-attachments/assets/7102f46f-ad04-4d2f-99b7-50fcd8e7ab28)

ITaluone commented:
Ah.. that's the issue.. it is essential to have a master/main branch so this default gitversion config can kick in:
```yml
main:
    label: ''
    increment: Patch
    prevent-increment:
      of-merged-branch: true
    track-merge-target: false
    track-merge-message: true
    regex: ^master$|^main$
    source-branches: []
    is-source-branch-for: []
    tracks-release-branches: false
    is-release-branch: false
    is-main-branch: true
    pre-release-weight: 55000
```

bprussell commented:
Success! I simply checked out the upstream master branch and pushed that to my fork, and checked out my feature branch again, and now the build script works.

Thank you for your help @ITaluone !

ITaluone commented:
Cool, glad to hear this :D

dennisdoomen commented:
> Yeah, looks like he used a special kind of clone that didn't include all the other branches and tags.

So that was the issue after all. I never noticed that "develop-only" option, so it's good to know that GitHub is suggesting this now. Happy to see that the community helped you out. Kudos to @ITaluone

IT-VBFK commented:
>> Yeah, looks like he used a special kind of clone that didn't include all the other branches and tags.

> So that was the issue after all.

Actually, he used a special kind of fork 😉

jnyrup commented on Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.HaveMilliseconds.cs:
```diff
@@ -0,0 +1,127 @@
+            subject.Should().HaveMilliseconds(expectation);
+        }
+
+        [Fact]
+        public void Different_value_throws()
+        {
+            // Arrange
+            DateTime subject = 31.December(2009).At(23, 59, 59, 999);
+            int expectation = 1;
+
+            // Act
+            Action act = () => subject.Should().HaveMilliseconds(expectation);
+
+            // Assert
+            act.Should().Throw<XunitException>()
+                .WithMessage("Expected the milliseconds part of subject to be 1, but found 999.");
+        }
+
+        [Fact]
+        public void Different_value_throws_with_reason()
+        {
+            // Arrange
+            DateTime subject = 31.December(2009).At(23, 59, 59, 999);
+            int expectation = 1;
+
+            // Act
+            Action act = () => subject.Should().HaveMilliseconds(expectation, "because we want to test the failure case with {0} and {1}", expectation, subject.Millisecond);
+
+            // Assert
+            act.Should().Throw<XunitException>()
+                .WithMessage("Expected the milliseconds part of subject to be 1 because we want to test the failure case with 1 and 999, but found 999.");
+        }
```
As `Different_value_throws_with_reason` covers the same subject+expectation combo as `Different_value_throws`, I would be fine just having that test instead of spread across two tests.

bprussell replied:
Sure, makes sense. Done.

bprussell commented:
@dennisdoomen Why was this closed?

dennisdoomen commented:
My bad. That happened accidentally because I deleted the develop branch. Contributions for 8.x can go directly to the main branch now. I forgot that deleting a branch also closes the PRs targeting that branch.

bprussell commented:
@dennisdoomen Ok, thanks, I will open a new one. However, the only thing holding this up, as far as I know, is that the issue doesn't have an `api-approved` label. I didn't know if I should poke someone about it, or just be patient and wait to see if it ever got added. I guess having the PR get deleted prompted me to poke you about it. :) Thanks.

dennisdoomen commented:
I think the only thing left was to re-review the rework, but we got occupies with v8 work.

I assume you are aware of the license changes that we introduced in v8, right?
